torch/csrc/generic/Tensor.cpp

```diff
@@ -496,14 +503,363 @@ static bool THPTensor_(_indexOnce)(PyObject *index, int &indexed_dim,
+    }
+    indexers.push_back(indexer);
+  }
+  Py_DECREF(fast);
+
+  THIndexTensor **maybeBroadcasted = (THIndexTensor **)THAlloc(seqCount * sizeof(THIndexTensor*));
+  THIndexTensor **candidates = (THIndexTensor **)THAlloc(seqCount * sizeof(THIndexTensor*));
+
+  for (Py_ssize_t i = 0; i < seqCount; ++i) {
+    maybeBroadcasted[i] = THIndexTensor_(new)(LIBRARY_STATE_NOARGS);
+    candidates[i] = THIndexTensor_(newWithTensor)(LIBRARY_STATE indexers[i]->cdata);
+  }
+
+  // Broadcast/Expand indexing Tensors as necessary
+  bool broadcastSuccess = true;
+  try {
+    THIndexTensor_(expandNd)(LIBRARY_STATE maybeBroadcasted, candidates, seqCount);
+    // Place Broadcasted Tensors into output vector, implicitly transferring
+    // ownership
+    for (Py_ssize_t i = 0; i < seqCount; ++i) {
+      broadcasted.push_back(maybeBroadcasted[i]);
+    }
+  } catch (std::exception& e) {
+    // Broadcasted failed, cleanup and set error
+    for (int i = 0; i < seqCount; ++i) {
+      THIndexTensor_(free)(LIBRARY_STATE maybeBroadcasted[i]);
+    }
+    broadcastSuccess = false;
+    PyErr_Format(PyExc_IndexError, "The advanced indexing objects could not be broadcast");
+  }
+
+  // No matter what, need to cleanup the candidates
+  for (Py_ssize_t i = 0; i < seqCount; ++i) {
+    THIndexTensor_(free)(LIBRARY_STATE candidates[i]);
+  }
+  THFree(candidates);
+  THFree(maybeBroadcasted);
+
+  return broadcastSuccess;
+}
+
+// Caller takes ownership of the returned IndexTensor
+static THIndexTensor* THPTensor_(_calculateLinearIndices)(
+    THTensorPtr& indexed, std::vector<THIndexTensor *>& broadcasted) {
+
+  // Get the number of indices to generate - this will be equal to the number
+  // of elements in each broadcasted Tensor
+  ptrdiff_t indexingElements = THIndexTensor_(nElement)(LIBRARY_STATE broadcasted.at(0));
+  THIndexTensor *linearIndices = THIndexTensor_(newWithSize1d)(LIBRARY_STATE indexingElements);
+  THLongStorage *indexerSize = THLongStorage_newWithSize(1);
+  THLongStorage_set(indexerSize, 0, indexingElements);
+
+  for (ptrdiff_t i = 0; i < indexingElements; ++i) {
+    long linearIdx = THTensor_(storageOffset)(LIBRARY_STATE indexed);
+    for (int j = broadcasted.size() - 1; j >= 0; --j) {
+      // Given a series of broadcasted Tensors, we take the jth tensor from the sequence (representing
+      // the indices at dim j) and grab the ith value (used in generating the ith output value)
+      THIndexTensor *indexer = THIndexTensor_(newContiguous)(LIBRARY_STATE broadcasted.at(j));
+
+      // The indexing tensor might not be one-dimensional, but we are generating a vector of
+      // indices, so we need to view the indexer as 1D prior to getting the value for the
+      // particular dimension
+      THIndexTensor *oned = THIndexTensor_(newView)(LIBRARY_STATE indexer, indexerSize);
+
+      // Actually laod the value, and verify it is not out-of-bounds
+      long indexAtDim = THIndexTensor_(get1d)(LIBRARY_STATE oned, i);
```

Comment:
In the short term to get this checked in, I may have to keep it all on the GPU. Unfortunately dealing with non-generic types in `generic/Tensor.cpp` is a little tricky. Ultimately, all the index calculation will be done on the GPU, and only once we are ready to do the indexSelect will we copy that indexing Tensor back to the GPU.

In the long term, index calculation could be a kernel in THC.